src/libraries/Microsoft.Extensions.Caching.Abstractions/src/MemoryCacheExtensions.cs

```diff
@@ -198,5 +225,32 @@ public static TItem Set<TItem>(this IMemoryCache cache, object key, TItem value,
 
             return (TItem?)result;
         }
+
+        /// <summary>
+        /// Asynchronously gets the value associated with this key if it exists, or generates a new entry using the provided key and a value from the given factory if the key is not found.
+        /// </summary>
+        /// <typeparam name="TItem">The type of the object to get.</typeparam>
+        /// <param name="cache">The <see cref="IMemoryCache"/> instance this method extends.</param>
+        /// <param name="key">The key of the entry to look for or create.</param>
+        /// <param name="factory">The factory task that creates the value associated with this key if the key does not exist in the cache.</param>
+        /// <param name="createOptions">The <see cref="MemoryCacheEntryOptions"/> needs to be applied to the created <see cref="ICacheEntry"/> if not null.</param>
+        /// <returns>The task object representing the asynchronous operation.</returns>
+        public static async Task<TItem?> GetOrCreateAsync<TItem>(this IMemoryCache cache, object key, Func<ICacheEntry, Task<TItem>> factory, MemoryCacheEntryOptions? createOptions)
+        {
+            if (!cache.TryGetValue(key, out object? result))
+            {
+                using ICacheEntry entry = cache.CreateEntry(key);
+
+                if (createOptions != null)
+                {
+                    entry.SetOptions(createOptions);
+                }
+
+                result = await factory(entry).ConfigureAwait(false);
```

Comment:
If `GetOrCreateAsync` is called again while `await factory(entry)` is still in progress, this method will return `null`. I think the await needs to be moved above `CreateEntry` to avoid this

Reply:
```c#
await using var services = new ServiceCollection()
    .AddMemoryCache()
    .BuildServiceProvider();
var memoryCache = services.GetRequiredService<IMemoryCache>();

Task.Run(() =>
{
    memoryCache.GetOrCreate("test", entry =>
    {
        Thread.Sleep(4000);
        Console.WriteLine("completed");
        return "1213";
    });
});
Thread.Sleep(2000);
if (memoryCache.TryGetValue("test", out var value))
{
    Console.WriteLine($"cache get and value is {value}");
}
else
{
    Console.WriteLine("cache not exists");
}
```

tried with a small demo, got `cache not exists` output, the cache entry would not be committed to the memory cache entries when it created